### PR TITLE
3D point selection with drag box

### PR DESCRIPTION
**Pull request by kevinyamauchi**
# Description
This PR adds the ability to perform selection of points in 3D with a drag box. The selection algorithm works as follows:

1. Project the points in view and drag box corners on to a plane parallel to the canvas (i.e., normal direction is the view direction).
2. Rotate the points and drag box such that the drag box is axis aligned
3. Since the points and drag box lie in the same plane, we can determine which points are in the bounding box in 2D .

## Type of change
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Supercedes https://github.com/napari/napari/pull/3593

# How has this been tested?
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by kevinyamauchi** [comment]
Here is a video of this working:

https://user-images.githubusercontent.com/1120672/146575083-8c5c7d71-da35-421a-90a7-9ef280cac7a1.mov

**Comment by brisvag** [comment]
Thsi si amazing!

**Comment by kevinyamauchi** [comment]
The test failure appears unrelated. ~~I think vispy must have changed their `get_colormap()` function to no longer accept `'gray'`.~~ edit: we were no longer installing matplotlib in our test env.

<details>
  <summary>Error message</summary>

```python
  
      attribute = 'edge'
  
      @pytest.mark.parametrize("attribute", ['edge', 'face'])
      def test_add_colormap(attribute):
          """Test  directly adding a vispy Colormap object"""
          shape = (10, 2)
          np.random.seed(0)
          data = 20 * np.random.random(shape)
          annotations = {'point_type': _make_cycled_properties([0, 1.5], shape[0])}
          color_kwarg = f'{attribute}_color'
          colormap_kwarg = f'{attribute}_colormap'
          args = {color_kwarg: 'point_type', colormap_kwarg: 'viridis'}
          layer = Points(data, properties=annotations, **args)
      
  >       setattr(layer, f'{attribute}_colormap', get_colormap('gray'))
  
  napari/layers/points/_tests/test_points.py:973: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  name = 'gray', args = (), kwargs = {}
  
      def get_colormap(name, *args, **kwargs):
          """Obtain a colormap.
      
          Parameters
          ----------
          name : str | Colormap
              Colormap name. Can also be a Colormap for pass-through.
          *args:
              Deprecated.
          **kwargs
              Deprecated.
      
          Examples
          --------
          >>> get_colormap('autumn')
          >>> get_colormap('single_hue')
      
          .. versionchanged: 0.7
      
              Additional args/kwargs are no longer accepted. Colormap classes are
              no longer created on the fly. To create a ``cubehelix``
              (``CubeHelixColormap``), ``single_hue`` (``SingleHue``), ``hsl``
              (``HSL``), ``husl`` (``HSLuv``), ``diverging`` (``Diverging``), or
              ``RdYeBuCy`` (``RedYellowBlueCyan``) colormap you must import and
              instantiate it directly from the ``vispy.color.colormap`` module.
      
          """
          if args or kwargs:
              warnings.warn("Creating a Colormap instance with 'get_colormap' is "
                            "no longer supported. No additional arguments or "
                            "keyword arguments should be passed.", DeprecationWarning)
          if isinstance(name, BaseColormap):
              return name
      
          if not isinstance(name, str):
              raise TypeError('colormap must be a Colormap or string name')
          if name in _colormaps:  # vispy cmap
              cmap = _colormaps[name]
              if name in ("cubehelix", "single_hue", "hsl", "husl", "diverging", "RdYeBuCy"):
                  warnings.warn(
                      f"Colormap '{name}' has been deprecated since vispy 0.7. "
                      f"Please import and create 'vispy.color.colormap.{cmap.__class__.__name__}' "
                      "directly instead.",
                      DeprecationWarning,
                      stacklevel=2,
                  )
      
          elif has_matplotlib():  # matplotlib cmap
              try:
                  cmap = MatplotlibColormap(name)
              except ValueError:
                  raise KeyError('colormap name %s not found' % name)
          else:
  >           raise KeyError('colormap name %s not found' % name)
  E           KeyError: 'colormap name gray not found'
```
</details>

Edit: it looks like `grays` would be accepted. Alternatively if matplotlib is installed, `gray` should work.

**Comment by brisvag** [comment]
>     1. Project the points in view and drag box corners on to a plane parallel to the canvas (i.e., normal direction is the view direction).
> 
>     2. Rotate the points and drag box such that the drag box is axis aligned
> 
>     3. Since the points and drag box lie in the same plane, we can determine which points are in the bounding box in 2D .

Can you elaborate on `2.` a bit?

**Comment by kevinyamauchi** [comment]
>>     1. Project the points in view and drag box corners on to a plane parallel to the canvas (i.e., normal direction is the view direction).
>> 
>>     2. Rotate the points and drag box such that the drag box is axis aligned
>> 
>>     3. Since the points and drag box lie in the same plane, we can determine which points are in the bounding box in 2D .
>
>Can you elaborate on `2.` a bit?

We rotate the points/bbox corners to a new basis comprising the bbox normal, the camera up direction, and the "horizontal direction" (i.e., vector orthogonal to bbox normal and camera up direction). This makes it such that the bounding box is axis aligned (i.e., the major and minor axes of the bounding box are aligned with the new 0 and 1 axes).

https://github.com/kevinyamauchi/napari/blob/36190aa64b5781f3fc4f551860656e989c519b7b/napari/layers/points/_points_utils.py#L144-L154

**Comment by tlambert03** [comment]
> Edit: it looks like grays would be accepted. Alternatively if matplotlib is installed, gray should work.

merge in #3837 and it should work

**Comment by kevinyamauchi** [comment]
Will do. Thanks, @tlambert03 !

**Comment by kevinyamauchi** [comment]
I think this is ready for a first round of review! cc @alisterburt , @brisvag

File: napari/layers/base/base.py [hunk 1]
```diff
@@ -1295,6 +1291,38 @@ def _world_to_displayed_data_ray(
...
+    def _world_to_data_dims_displayed(
+        self, dims_displayed: List[int], ndim_world: int
+    ) -> np.ndarray:
+        """Convert the displayed dims in world coordinates to the data coordinates.
+
+        This accounts for differences in dimensionality of the world and the data.
+        for example a global order of [2, 1, 0, 3] -> [0, 1] for a layer that
+        only has two dimensions or -> [1, 0, 2] for a layer with three as that
+        corresponds to the relative order of the last two and three dimensions
+        respectively
```
Comment (alisterburt) [inline napari/layers/base/base.py]: I think I understand the need for this but I found this docstring quite hard to parse - I don't think I understand it well enough to have a better suggestion so don't consider it blocking...

Am I right in thinking that this is needed only when there is both a mix in dimensionality and the dims have been rolled?

File: napari/layers/points/_points_mouse_bindings.py [hunk 2]
```diff
@@ -76,22 +80,17 @@ def select(layer, event):
...
-                layer.selected_data = list(target)
-            else:
-                layer.selected_data = layer._indices_view[selection]
-        else:
-            layer.selected_data = set()
+        n_display = len(event.dims_displayed)
+        _select_points_from_drag(
+            layer=layer, modify_selection=modify_selection, n_display=n_display
+        )
+
```
Comment (alisterburt) [inline napari/layers/points/_points_mouse_bindings.py:90]: this is looking clean!

File: napari/layers/points/_points_mouse_bindings.py [hunk 3]
```diff
@@ -139,3 +138,83 @@ def _toggle_selected(selected_data, value):
...
+        ndim_world = len(event.position)
+        # if in 3D, set the drag normal and up directions
+        # get the indices of the displayed dimensions
+        dims_displayed_data = layer._world_to_data_dims_displayed(
+            dims_displayed=event.dims_displayed, ndim_world=ndim_world
+        )
+
+        # get the view direction in data coordinates
+        layer._drag_normal = np.asarray(
+            layer._world_to_data_ray(event.view_direction)
```
Comment (alisterburt) [inline napari/layers/points/_points_mouse_bindings.py]: could use `layer._world_to_displayed_data_ray()` here

File: napari/layers/points/_points_mouse_bindings.py [hunk 4]
```diff
@@ -139,3 +138,83 @@ def _toggle_selected(selected_data, value):
...
+        )[dims_displayed_data]
+
+        # get the up direction of the camera
+        layer._drag_up = np.asarray(
+            layer._world_to_data_ray(event.up_direction)
+        )[dims_displayed_data]
```
Comment (alisterburt) [inline napari/layers/points/_points_mouse_bindings.py]: minor nitpick, maybe packing this into a util function would increase readability?

File: napari/layers/points/_points_mouse_bindings.py [hunk 5]
```diff
@@ -139,3 +138,83 @@ def _toggle_selected(selected_data, value):
...
+    layer : napari.layers.Points
+        The points layer to select points on.
+    modify_selection : bool
+        Set to true if the selection should modify the current selected data
+        in layer.selected_data.
+    n_display : int
+        The number of dimensions current being displayed
+    """
+    if len(layer._view_data) > 0:
+        # if there is data in view, find the points in the drag box
```
Comment (alisterburt) [inline napari/layers/points/_points_mouse_bindings.py]: you could drop a level of nesting here by doing 

```suggestion
    if len(layer._view_data) > 0:
        layer.selected_data = set()
        return
```

I find the above a little more readable, keeps the code specific to each case a bit more together

File: napari/layers/points/_points_utils.py [hunk 6]
```diff
@@ -1,10 +1,51 @@
...
 
+from ...utils.geometry import project_points_onto_plane
 from ...utils.translations import trans
 
 
+def _create_box_from_corners_3d(
+    box_corners: np.ndarray, box_normal: np.ndarray, up_vector: np.ndarray
+) -> np.ndarray:
+    """Get the corners for a box in 3D from the corners, the normal direction,
+    and the up direction.
```
Comment (alisterburt) [inline napari/layers/points/_points_utils.py]: I'm finding it a little hard to grasp exactly what this method does, is it finding the corner positions for a 2d cuboid in 3D? These geometry things are always weird to describe

File: napari/layers/points/_points_utils.py [hunk 7]
```diff
@@ -45,17 +86,85 @@ def points_to_squares(points, sizes):
...
     rect = np.concatenate(
         [
-            points + np.sqrt(2) / 2 * np.array([sizes, sizes]).T,
-            points + np.sqrt(2) / 2 * np.array([sizes, -sizes]).T,
-            points + np.sqrt(2) / 2 * np.array([-sizes, sizes]).T,
-            points + np.sqrt(2) / 2 * np.array([-sizes, -sizes]).T,
+            points + 0.5 * np.array([sizes, sizes]).T,
+            points + 0.5 * np.array([sizes, -sizes]).T,
+            points + 0.5 * np.array([-sizes, sizes]).T,
+            points + 0.5 * np.array([-sizes, -sizes]).T,
```
Comment (alisterburt) [inline napari/layers/points/_points_utils.py:95]: I don't see where `sizes` is defined here, am I missing something?

File: napari/layers/points/_points_utils.py [hunk 8]
```diff
@@ -45,17 +86,85 @@ def points_to_squares(points, sizes):
...
-def points_in_box(corners, points, sizes):
+def _points_in_box_3d(
+    box_corners: np.ndarray,
+    points: np.ndarray,
+    sizes: np.ndarray,
+    box_normal: np.ndarray,
+    up_direction: np.ndarray,
+) -> List[int]:
+    """Determine which points are inside of a by a bounding box that is extended
+    along its normal direction.
```
Comment (alisterburt) [inline napari/layers/points/_points_utils.py]: This might be easier to understand if described in terms of 2D view of a 3D scene? Something like
```suggestion
    """Determine which points lie inside of a 2D bounding box when viewed from a specific direction.
```

**Review by alisterburt — COMMENTED** [review]
This is looking great @kevinyamauchi !! I've left some minor comments on the code but nothing crazy, just things where some more explanation would be useful. Really excited to get this in, happy to help you push it over the line over the break if you like. Test coverage is really nice :)

Something doesn't seem to be working 100% properly when moving multiple points in the nD_points example, any ideas what could be going on here? I also get a weird 'snap to somewhere else in the dims' after moving the points

https://user-images.githubusercontent.com/7307488/146651982-62d2d090-86cd-4ebd-82be-f76a2a730d94.mp4

File: napari/layers/points/_points_utils.py (hunk 7 above)
Comment (kevinyamauchi) [reply napari/layers/points/_points_utils.py:95]: Ah yeah, it is a little confusing how github displayed the diff. This is from the function `points_to_squares()`, not `create_box()`. It just looks like it is `create_box()` because the `def points_to_squares()` part is folded. You can either view the whole file or click the unfold button on the left side to see the whole function.

File: napari/layers/base/base.py (hunk 1 above)
Comment (kevinyamauchi) [reply napari/layers/base/base.py]: Yes, this is for determining which dimensions of the layer data are being displayed. It is most important for the case when the layer has fewer dimensions than the world, but it also works when layer ndim is the same as world ndim. I can try to rework the docstring.

**Comment by kevinyamauchi** [comment]
Thanks for the review, @alisterburt! 

> Something doesn't seem to be working 100% properly when moving multiple points in the nD_points example, any ideas what could be going on here? I also get a weird 'snap to somewhere else in the dims' after moving the points

I think this is due to the dtype of the points being an integer type, so they there is some discretization error. I don't see the behavior you showed when I set the point coordinate dtype to float (see example below. Does that fix it for you as well?

```python
"""
Display one points layer on top of one 4-D image layer using the
add_points and add_image APIs, where the markes are visible as nD objects
across the dimensions, specified by their size
"""

import numpy as np
from skimage import data
import napari


blobs = np.stack(
    [
        data.binary_blobs(
            length=128, blob_size_fraction=0.05, n_dim=3, volume_fraction=f
        )
        for f in np.linspace(0.05, 0.5, 10)
    ],
    axis=0,
)
viewer = napari.view_image(blobs.astype(float))

# add the points
points = np.array(
    [
        [0, 0, 100, 100],
        [0, 0, 50, 120],
        [1, 0, 100, 40],
        [2, 10, 110, 100],
        [9, 8, 80, 100],
    ], dtype=float
)
viewer.add_points(
    points, size=[0, 6, 10, 10], face_color='blue', n_dimensional=True
)

napari.run()
```

File: napari/layers/points/_points_mouse_bindings.py (hunk 5 above)
Comment (jni) [reply napari/layers/points/_points_mouse_bindings.py]: ... should that be a `== 0`, @alisterburt?

Comment (kevinyamauchi) [reply napari/layers/points/_points_mouse_bindings.py]: Good point @alisterburt . I've implemented it as `if len(layer._view_data) == 0` as @jni suggested.

File: napari/layers/points/_points_utils.py (hunk 6 above)
Comment (kevinyamauchi) [reply napari/layers/points/_points_utils.py]: I've updated the docstring as below. Does this seem more clear?

```
    """Get the corners for a box in 3D from two opposing corners, the normal direction,
    and the up direction.

    The resulting box will include the two corners passed in as
    box_corners, lie in a plane with normal box_normal, and have one of its
    axes aligned with the up_vector.

    Parameters
    ----------
    """"

File: napari/layers/base/base.py (hunk 1 above)
Comment (kevinyamauchi) [reply napari/layers/base/base.py]: I've updated the docstring a bit. Let me know if there are specific things that are still unclear.

File: napari/layers/points/_points_mouse_bindings.py (hunk 3 above)
Comment (kevinyamauchi) [reply napari/layers/points/_points_mouse_bindings.py]: Done!

File: napari/layers/base/base.py (hunk 1 above)
Comment (alisterburt) [reply napari/layers/base/base.py]: looks great! Great job figuring this out, my interpretation of how this *should* work was actually wrong... it's worth noting that this behaviour may change if we move forward with changes to how dims are handled discussed in #3848

File: napari/layers/points/_points_utils.py (hunk 6 above)
Comment (alisterburt) [reply napari/layers/points/_points_utils.py]: Much clearer, thanks!

File: napari/layers/points/_points_utils.py (hunk 7 above)
Comment (alisterburt) [reply napari/layers/points/_points_utils.py:95]: my bad!

**Review by alisterburt — COMMENTED** [review]
This is looking very close! Good catch with the dtype in the example, I forgot we had come across that 

I'm still getting an undesired change in `Dims.current_step` after moving points in the `nD_points` example

https://user-images.githubusercontent.com/7307488/147076403-06930fbd-53a8-43ff-b2cd-be8b4e96572d.mp4

I think what's going on is that the data change event is triggering a dims change, I'm looking into it now, once this is fixed I think we can merge!

File: napari/_vispy/layers/points.py [hunk 9]
```diff
@@ -88,27 +88,22 @@ def _on_highlight_change(self):
...
-            else:
-                pos = self.layer._highlight_box
-                width = settings.appearance.highlight_thickness
-
-            self.node._subvisuals[2].set_data(
-                pos=pos[:, ::-1],
-                color=self._highlight_color,
-                width=width,
-            )
+        # if self.layer._ndisplay == 2:
```
Comment (brisvag) [inline napari/_vispy/layers/points.py]: Cooments are probably obsolete.

File: napari/layers/base/base.py [hunk 10]
```diff
@@ -1405,7 +1434,7 @@ def _get_ray_intersections(
         world : bool
             True if the provided coordinates are in world coordinates.
             Default value is True.
-        bounding_box : np.ndarray
+        bounding_box: np.ndarray
```
Comment (brisvag) [inline napari/layers/base/base.py]: Are these colons supposed to always have a space on both sides or is this ok?

**Comment by alisterburt** [comment]
The data change event triggered here
https://github.com/napari/napari/blob/55f2d7ef929762bd99412d854236fe0649d4f359/napari/layers/points/points.py#L1652
leads to the following, ending with the dims being changed
image

**Comment by alisterburt** [comment]
confirmed that this is also a problem in 2D and should be fixed separately, will raise a separate issue - I think we're good to merge here!

File: napari/layers/points/_points_utils.py [hunk 11]
```diff
@@ -1,10 +1,55 @@
...
+def _create_box_from_corners_3d(
+    box_corners: np.ndarray, box_normal: np.ndarray, up_vector: np.ndarray
+) -> np.ndarray:
+    """Get the corners for a box in 3D from two opposing corners, the normal direction,
+    and the up direction.
+
+    The resulting box will include the two corners passed in as box_corners,
+    lie in a plane with normal box_normal, and have one of its axes aligned
+    with the up_vector.
+
```
Comment (brisvag) [inline napari/layers/points/_points_utils.py]: Super clear, love this!

File: napari/layers/points/_points_utils.py (hunk 8 above)
Comment (brisvag) [reply napari/layers/points/_points_utils.py]: I think this is still a bit unclear. Something that should be clarified is that the box is not a parallelepiped, but it has an infinite "side" in the direction of `box_normal`... or if I got it wrong, then it needs to be clarifed otherwise :P

File: napari/layers/points/points.py [hunk 12]
```diff
@@ -1624,8 +1636,17 @@ def remove_selected(self):
...
-            self.selected_data = set()
+            else:
+                if self._value is not None:
+                    # update the index of self._value to account for the
+                    # data being removed
+                    indices_removed = np.array(index) < self._value
+                    offset = np.sum(indices_removed)
+                    self._value -= offset
+                    self._value_stored -= offset
+
```
Comment (brisvag) [inline napari/layers/points/points.py:1647]: Was this a bug getting fixed, or is it more related with 3D specifically than it seems? If yes, how?

**Review by brisvag — APPROVED** [review]
A few minor comments, mostly about inline comments and docstrings. Looks amazing @kevinyamauchi!

**Comment by brisvag** [comment]
> confirmed that this is also a problem in 2D and should be fixed separately, will raise a separate issue - I think we're good to merge here!

@alisterburt I think this was already addressed in #3859, after this self-centering mechanism was was introduced in #3718.

EDIT: I can confirm that merging main here will solve this.

**Review by alisterburt — APPROVED** [review]

File: napari/_vispy/layers/points.py (hunk 9 above)
Comment (Carreau) [inline napari/_vispy/layers/points.py]:
```suggestion
```

File: napari/components/camera.py [hunk 13]
```diff
@@ -162,3 +162,28 @@ def calculate_nd_view_direction(
         view_direction_nd = np.zeros(ndim)
         view_direction_nd[list(dims_displayed)] = self.view_direction
         return view_direction_nd
+
+    def calculate_nd_up_direction(
+        self, ndim: int, dims_displayed: Tuple[int]
+    ) -> np.ndarray:
```
Comment (Carreau) [inline napari/components/camera.py]:
```suggestion
    ) -> Optional[np.ndarray]:
```

File: napari/layers/base/base.py [hunk 14]
```diff
@@ -1295,6 +1291,39 @@ def _world_to_displayed_data_ray(
...
+        -------
+        dims_displayed_data : np.ndarray
+            The displayed dimensions in data coordinates.
+        """
+        offset = ndim_world - self.ndim
+        order = np.array(dims_displayed)
+        if offset <= 0:
+            dims_displayed_data = list(range(-offset)) + list(order - offset)
+        else:
+            dims_displayed_data = list(order[order >= offset] - offset)
```
Comment (Carreau) [inline napari/layers/base/base.py]: In the sig and docstring you say you return an ndarray,  but here it's a list. And I guess you can return from within the if/else directly ? I'm not sure if it's cleaner or not.

File: napari/layers/base/base.py (hunk 10 above)
Comment (Carreau) [reply napari/layers/base/base.py]: Yeah, if there is no space it's missparsed by numpydoc.
```suggestion
        bounding_box : np.ndarray
```

File: napari/layers/points/_points_utils.py [hunk 15]
```diff
@@ -45,17 +90,85 @@ def points_to_squares(points, sizes):
...
+    points_axis_aligned = projected_points @ plane_basis
+    points_axis_aligned = points_axis_aligned[:, :2]
+
+    # determine which points are in the box using the
+    # axis-aligned basis
+    inside = points_in_box(
+        bbox_corners_axis_aligned, points_axis_aligned, sizes
+    )
+
+    return inside
```
Comment (Carreau) [inline napari/layers/points/_points_utils.py]:
```suggestion
    return points_in_box(
        bbox_corners_axis_aligned, points_axis_aligned, sizes
    )
```

Up to you.

File: napari/layers/points/points.py (hunk 12 above)
Comment (kevinyamauchi) [reply napari/layers/points/points.py:1647]: Yeah, this was a fix for an unrelated bug that I caught while testing. The bug was:

1. Points selected (2D or 3D)
2. A point outside of the selection is highlighted (i.e., the mouse is hovered over it).
3. Delete the points while still hovering over the point
4. Since the index of the highlighted point is not updated, an `IndexError` is raised if that point had an index value greater than some of those that were deleted.

File: napari/layers/points/_points_utils.py (hunk 8 above)
Comment (kevinyamauchi) [reply napari/layers/points/_points_utils.py]: Sorry, I hadn't fixed it yet. The box is actually "infinite" in both directions along the `box_normal`. I will update the docstring now.

File: napari/_vispy/layers/points.py (hunk 9 above)
Comment (kevinyamauchi) [reply napari/_vispy/layers/points.py]: Fixed. Thanks!

File: napari/layers/base/base.py (hunk 10 above)
Comment (kevinyamauchi) [reply napari/layers/base/base.py]: Good catch. Thanks!

File: napari/_vispy/layers/points.py (hunk 9 above)
Comment (kevinyamauchi) [reply napari/_vispy/layers/points.py]: Fixed, thanks!

File: napari/layers/base/base.py (hunk 14 above)
Comment (kevinyamauchi) [reply napari/layers/base/base.py]: Good catch. I've updated the docstring and now return from within the if/else.

File: napari/layers/points/_points_utils.py (hunk 8 above)
Comment (kevinyamauchi) [reply napari/layers/points/_points_utils.py]: I've now updated the docstring. Let me know if you think it should be further clarified, @brisvag and @alisterburt !

**Comment by kevinyamauchi** [comment]
Thank you for the reviews, @alisterburt, @brisvag , and @Carreau ! I believe I have addressed your comments.

**Review by jni — APPROVED** [review]
Just fixed some docstrings as I went along, but everything is good! :smiley: thanks everyone for all the review till now!

**Review by brisvag — APPROVED** [review]
Yup, all good for me! Great PR @kevinyamauchi :D

File: napari/layers/base/base.py [hunk 16]
```diff
@@ -1295,6 +1291,36 @@ def _world_to_displayed_data_ray(
         vector_data_ndisplay /= np.linalg.norm(vector_data_ndisplay)
         return vector_data_ndisplay
 
+    def _world_to_data_dims_displayed(
+        self, dims_displayed: List[int], ndim_world: int
+    ) -> List[int]:
```
Comment (alisterburt) [inline napari/layers/base/base.py:1296]: not blocking but should we consider renaming this to `_world_to_displayed_data_dims` be more consistent with other 'world to scene' type methods before merge?

https://github.com/napari/napari/blob/1f68ec2022c35f661fb2e4a704a67925aabb5ba8/napari/layers/base/base.py#L1218-L1239

Comment (kevinyamauchi) [reply napari/layers/base/base.py:1296]: Good question. My interpretation is the naming convention is `<starting coord system>_<ending coord system>_<thing being transformed>`. For example, in the method you linked to above, it takes coordinates from world to the displayed data coordinates. Similarly, `layer.world_to_data_ray` takes a ray from world to data coordinates. In that frame, I named this method `_world_to_data_dims_displayed` since it is taking `dims_displayed` from world to data coordinates. Note that this is not "displayed data" coordinates referenced in `_world_to_displayed_data()`. 

That being said, I don't feel strongly about the name. What do you think? If it wasn't clear that the function is taking `dims_displayed` from world to data coordinates, perhaps I should refine the docstring.

Comment (alisterburt) [reply napari/layers/base/base.py:1296]: discussed offline - let's leave this as is!

**Review by sofroniewn — COMMENTED** [review]
Looks like this is ready to go in! Thanks for the reviews all. I'll give @kevinyamauchi a chance to weigh in once more, but if i don't hear anything in a bit then i think we can merge 🚀

File: napari/layers/base/base.py (hunk 16 above)
Comment (brisvag) [reply napari/layers/base/base.py:1296]: > In that frame, I named this method `_world_to_data_dims_displayed` since it is taking `dims_displayed` from world to data coordinates. Note that this is not "displayed data" coordinates referenced in `_world_to_displayed_data()`.

Wow, things are getting messy :D

Comment (kevinyamauchi) [reply napari/layers/base/base.py:1296]: Yeah, we can definitely improve the naming convention a bit. Happy to hear suggestions. Since this is consistent with the rest of the codebase, I would suggest we proceed as is and then rename everything once the nomenclature is decided in #3848.

**Comment by kevinyamauchi** [comment]
> Looks like this is ready to go in! Thanks for the reviews all. I'll give @kevinyamauchi a chance to weigh in once more, but if i don't hear anything in a bit then i think we can merge 🚀

Nothing further from my side. I think it is good to go if there are no additional comments!

**Comment by alisterburt** [comment]
merging - congrats @kevinyamauchi, this feels like a huge step forwards! 🚀 🎉
